### DISCUSSION TITLE
Problem with RedirectResponse

### DISCUSSION THREAD

**Discussion by frchalaoux** [fastapi/fastapi#10408] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@router.delete("/delete/{id}")
def delete_listening(id: int,db: Session = Depends(get_db)):
    listening = retrieve_listening(id =id,db=db)
    if not listening:
        return HTTPException(status_code=status.HTTP_404_NOT_FOUND,detail=f"Listening with {id} does not exist")
    delete_listening_by_id(id=id,db=db)
    #return {"msg":"Successfully deleted."}
    #return {"detail": "Successfully deleted."}  #add this line
    # return templates.TemplateResponse(
    # "listenings/show_listenings_to_delete.html", {"request": request, "listenings": listenings}
    # )
    return RedirectResponse("/delete-listening-possibles/")  # Redirige vers la page de liste

@router.get("/delete-listening-possibles/")
def show_listenings_to_delete(request: Request, db: Session = Depends(get_db)):
    listenings = list_listenings(db=db)
    return templates.TemplateResponse(
        "listenings/show_listenings_to_delete.html", {"request": request, "listenings": listenings}
    )

#show_listenings_to_delete.html
<script type="text/javascript">
    function delete_listening(id){
        fetch('/listenings/delete/'+id,{
            method:'DELETE',})
        .then(response => response.json())
        .then(document.getElementById('result').innerHTML = "Refreshing...")
        .then(data => document.getElementById('result').innerHTML = data.detail);
    }
</script>

#HTML header from show_listenings_to_delete.html
<meta http-equiv="cache-control" content="no-cache, no-store, must-revalidate">
<meta http-equiv="pragma" content="no-cache">
<meta http-equiv="expires" content="0">
```

### Description

Hi all,

I want to suppress "listenings" in my DB and below this is the equivalent html table in show_listenings_to_delete.html. 

Id | Date de création | Status | Rapport |  
-- | -- | -- | -- | --
1 | 2023-10-06 00:00:00 | True | True | Delete button
2 | 2023-10-06 00:00:00 | True | True | Delete button
3 | 2023-10-06 00:00:00 | True | True | Delete button

When I click on a "Delete button", the record deletion is initialized with the javascript function delete_listening and an error appears : 

""DELETE /delete-listening-possibles/ HTTP/1.1" 405 Method Not Allowed"

My DB is updated but the page is not refreshed correctly with : 

"return RedirectResponse("/delete-listening-possibles/")" from @router.delete("/delete/{id}")

It seems that return RedirectResponse("/delete-listening-possibles/") does not display a correct state the DB while the DB is correct and the record deleted. I checked the routes and the endpoint and it seems alright. I tried also to change the operation

Any suggestion ?

### Operating System

macOS

### Operating System Details

machos = Ventura 13.6

### FastAPI Version

0.103.2

### Pydantic Version

2.4.2

### Python Version

Python 3.11.5

**Comment by jgould22** [comment, marked as answer]
You are using a redirect response which by default returns an [HTTP status code 307](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/307).

When the browser gets this code it will retry the request it just got a response for at the location specified in the `location` header.

So what is happening is you are requesting `"/delete/{id}"` with the DELETE method and then returning a 307 which is telling the browser don't make your request here do it at this location. So it is re-sending the same request to `/delete-listening-possibles/` . Since that router is only defined with a GET it is returning a 405 to you telling you that DELETE doesnt work here.

**Reply by frchalaoux** [reply]
Ok, thanks. What could be the solution ?

**Reply by jgould22** [reply]
I am not too much of a front-end dev but usually I think you would make your request in javascript to your delete endpoint and on success navigate to your second page.

Don't return a redirect from `/delete/{id}`, instead return a 200 success and then in javascript use something like 
```javascript
window.location.assign(
  "/delete-listening-possibles/",
);
```

or just grabbing the response with another request after your delete request

something like (my javascript is really rusty so consider this more pseudo code )
```javascript
#show_listenings_to_delete.html
<script type="text/javascript">
    function delete_listening(id){
        fetch('/listenings/delete/'+id,{
            method:'DELETE',})
        .then(document.getElementById('result').innerHTML = "Refreshing...")
        .then(function() {
               fetch('/delete-listening-possibles/')
                    .then(data => document.getElementById('result').innerHTML = data.detail);
         })
    }
</script>

**Reply by frchalaoux** [reply]
@jgould22, Thanks a lot for you help. I finally found a solution around your advice. Il seems to work correctly but ... not as I expected. But the important is to succeed, after ... Pyscript and more FastAPI knowledge would brings me a more elegant solution. If anybody else have cleaner solution, you are welcome. Read you Guys

The actual solution:

```
@router.delete("/delete/{id}")
def delete_listening(id: int, db: Session = Depends(get_db)):
    listening = retrieve_listening(id=id, db=db)
    if not listening:
        return HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=f"Listening with {id} does not exist")
    delete_listening_by_id(id=id, db=db)
    return {"redirect_url": "/delete-listening-possibles/"}  # Ajoutez cette ligne
```
```
function delete_listening(id) {
      fetch('/listenings/delete/' + id, {
          method: 'DELETE',
      })
      .then(response => response.json())
      .then(data => {
          if (data.redirect_url) {
              window.location.href = data.redirect_url;
          } else {
              document.getElementById('result').innerHTML = "Successfully deleted.";
          }
      });
  }
```
